argo-workflows/nautobot-interface-sync/code/sync_nautobot_interfaces.py

```diff
@@ -0,0 +1,58 @@
+import sushy
+from models import Chassis
+from helpers import *
+
+logger = setup_logger(__name__)
+
+
+def create_or_update_nb_intf(nb_session, nb_device, redfish_intf):
+    inf = nb_session.dcim.interfaces.get(device_id=nb_device.id, name=redfish_intf.name)
+    if inf is None:
+        logger.info(
+            f"{redfish_intf.name} was NOT found for {nb_device.name}, creating..."
+        )
+
+        nb_session.dcim.interfaces.create(
+            device=nb_device.id,
+            name=redfish_intf.name,
+            mac_address=redfish_intf.mac_addr,
+            type="other",
+            status="Active",
+            description=f"Location: {redfish_intf.location}",
+        )
+        logger.info(f"{redfish_intf.name} successfully created")
+    else:
+        logger.info(f"Found {inf.name} for {nb_device.name}")
+        if not inf.mac_address:
+            inf.update({"mac_address": i.mac_addr})
+            logger.info(f"Mac address for {inf.name} successfully updated")
+
+
+def main():
+    parser = get_parser(__file__)
+    args = parser.parse_args()
+
+    device_name = args.hostname
+    nb_url = args.nautobot_url or "http://nautobot-default.nautobot.svc.cluster.local"
+    nb_token = args.nautobot_token or get_cred("nb-token", "token")
+    oob_username = args.oob_username or get_cred("oob-secrets", "username")
+    oob_password = args.oob_password or get_cred("oob-secrets", "password")
+
+    pyn = get_nb_session(nb_url, nb_token)
+    device = get_nb_device(pyn, device_name)
+    oob_intf = get_nb_device_oob_interface(pyn, device)
+    oob_ip = get_ip_from_nb_interface(oob_intf)
+    oob = get_oob_sushy_session(oob_ip, oob_username, oob_password)
+
+    try:
+        chassis = Chassis.from_redfish(oob)
+    except sushy.exceptions.AccessError as e:
+        exit_with_error(e)
+
+    for intf in chassis.network_interfaces:
```

Comment:
is there any way of doing that update in a transaction, so that we don't get partial updates if any of them fail?

Reply:
adjusted